### ISSUE TITLE
hs100 platform won't load

### ISSUE THREAD

**Issue by ghope10** [plasticrake/homebridge-hs100#17]
I managed to get my homebridge working with my nest device, works great.
Now I am attempting to add my hs100 devices to homebridge. I have tried multiple ways to add the information to the config.json file but the platform will not load. It seems to still load the nest platform but homebridge doesn't seem to work.
I get the error:
```
[3/8/2017, 8:04:43 PM] Loaded plugin: homebridge-hs100
[3/8/2017, 8:04:43 PM] Registering platform 'homebridge-hs100.Hs100'
[3/8/2017, 8:04:43 PM] ---
[3/8/2017, 8:04:44 PM] Loaded plugin: homebridge-nest
[3/8/2017, 8:04:44 PM] Registering platform 'homebridge-nest.Nest'
[3/8/2017, 8:04:44 PM] ---
[3/8/2017, 8:04:44 PM] Loaded config.json with 0 accessories and 1 platforms.
[3/8/2017, 8:04:44 PM] ---
[3/8/2017, 8:04:44 PM] Loading 1 platforms...
[3/8/2017, 8:04:44 PM] Initializing Nest platform...
[3/8/2017, 8:04:44 PM] Fetching Nest devices.
Load homebridge-hs100.Hs100
/usr/local/lib/node_modules/homebridge-hs100/index.js:20
    this.plugs = config['plugs'] || [];
                       ^

TypeError: Cannot read property 'plugs' of null
    at new Hs100Platform (/usr/local/lib/node_modules/homebridge-hs100/index.js:20:24)
    at Server._loadDynamicPlatforms (/usr/local/lib/node_modules/homebridge/lib/server.js:315:30)
    at Server.run (/usr/local/lib/node_modules/homebridge/lib/server.js:82:8)
    at module.exports (/usr/local/lib/node_modules/homebridge/lib/cli.js:40:10)
    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:425:7)
    at startup (bootstrap_node.js:146:9)
    at bootstrap_node.js:540:3
```
Is this some incorrect parameters in the config file?

Sorry, new to this so any help would be much appreciated.
I can share the contents of the config file.

**Comment by SkyJohn** [comment]
I'm getting exactly the same error, not sure what is wrong.

**Comment by tucknology** [comment]
I also have the same error.

**Comment by dross199** [comment]
Getting the same error, have any of you guys made any progress at all ?

**Comment by tucknology** [comment]
There is a bug in version **2.0.0**, so I've installed version **1.1.1**. It is working perfectly now. The only downside is that the configuration requires that we manually specify the IP address of each switch. 

Install version 1.1.1 using the following command:

```
npm install -g homebridge-hs100@1.1.1
```

Then you'll need to add the following to your `config.json` file (see the old README.md)

```json
"platforms": [{
    "platform": "Hs100",
    "plugs": [{
        "host": "10.0.1.2",
        "port": 9999
    }, {
        "host": "10.0.1.3",
        "port": 9999
    }, {
        "host": "10.0.1.4",
        "port": 9999
    }]
}],
```

I found the IP address of my HS100 by looking through list of devices on my router. The port is always `9999`.

**Comment by dross199** [comment]
@tucknology Thanks for the Pointer but i am still getting the same error. TBH I am a complete noob when it comes to this, this is my first time looking at Node and PI's in general. I have put the error below: 

root@raspberrypi:~# homebridge
*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node&f=DNSServiceRegister>
[3/15/2017, 2:35:50 PM] Loaded plugin: homebridge-hs100
[3/15/2017, 2:35:50 PM] Registering platform 'homebridge-hs100.Hs100'
[3/15/2017, 2:35:50 PM] ---
[3/15/2017, 2:35:50 PM] config.json (/root/.homebridge/config.json) not found.
Load homebridge-hs100.Hs100
/usr/local/lib/node_modules/homebridge-hs100/index.js:20
    this.plugs = config['plugs'] || [];
                       ^

TypeError: Cannot read property 'plugs' of null
    at new Hs100Platform (/usr/local/lib/node_modules/homebridge-hs100/index.js:20:24)
    at Server._loadDynamicPlatforms (/usr/local/lib/node_modules/homebridge/lib/server.js:315:30)
    at Server.run (/usr/local/lib/node_modules/homebridge/lib/server.js:82:8)
    at module.exports (/usr/local/lib/node_modules/homebridge/lib/cli.js:40:10)
    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)

**Comment by dross199** [comment]
That has solved the issues but it is loading without any accessories any ideas ? My Json file is below:

{
    "bridge": {
        "name": "Homebridge",
        "username": "Removed",
        "port": 51826,
        "pin": "031-45-154"
    },

    "description": "This is an example configuration file. You can use this as a template for creating your own configuration file.",

    "platforms": [{
        "platform": "Hs100",
        "accessories": [{
            "accessory":"David Lamp",
            "host": "192.168.0.37",
            "port": 9999
        }]
    }],

    "accessories": []
}

**Comment by tucknology** [comment]
@dross199 You're not following the example config.json snippet correctly. My comments on your config are below:

```json
    "platforms": [{
        "platform": "Hs100",
        "accessories": [{     <<<< THIS SHOULD SAY PLUGS, NOT ACCESSORIES
             "accessory":"David Lamp",  <<< REMOVE THIS LINE
             "host": "192.168.0.37",
             "port": 9999
         }]
     }],

    "accessories": []   <<< REMOVE THIS
```

You should also check your config.json file using a tool like www.jsonlint.com to make sure you don't have any syntax errors.

**Comment by MrTeal** [comment]
@tucknology I tried your workaround but I am getting the following error message:
Homebridge is running on port 51826.
[3/16/2017, 8:32:05 PM] Done Configuring Cached Accessories: []
[3/16/2017, 8:32:08 PM] { [Error: connect EHOSTUNREACH 192.168.0.173:9999]
  code: 'EHOSTUNREACH',
  errno: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.0.173',
  port: 9999 }

I checked the IPs in my router and the plugs are working.
My config.json is as follows:
`
{
    "bridge": {
        "name": "HomePi",
        "username": "Removed",
        "port": 51826,
        "pin": "333-88-333"
    },

    "platforms": [
        {
            "platform": "Server",
            "port": 8765,
            "name": "Homebridge Server",
            "log": "/var/log/homebridge.log"
        },
        {
            "platform": "Hs100",
            "plugs": [{
        "host": "192.168.0.173",
        "port": 9999
    }, {
        "host": "192.168.0.165",
        "port": 9999
    }, {
        "host": "192.168.0.193",
        "port": 9999
    }]
        }
    ]
}
`

Any ideas? Thanks in advance!

**Comment by leematt** [comment]
This appears to be caused by not having the `Hs100` platform correctly configured in `config.json`. It should be in `platforms` in `config.json` like this:

```
"platforms": [
    {
      "platform": "Hs100"
    }
  ]
```

If it isn't properly configured, you'll get a null `config` value passed to the `Hs100` plugin, which then crashes when it tries to access `config['plugs']`.

If you can't get the configuration recognized for some reason, you could edit the `Hs100` plugin (`npm edit -g homebridge-hs100`) and remove the `this.plugs = config['plugs'] || [];` line ([#20](https://github.com/plasticrake/homebridge-hs100/blob/v2.0.0/index.js#L20)), as it appears to no longer be needed as of version `2.0.0`, which should also resolve this error. Or, just change `config['plugs']` to `this.config['plugs']`.

**Comment by MrTeal** [comment]
Hi @leematt , thanks for your response!
I edited the configuration to this:
```
{
    "bridge": {
        "name": "HomePi",
        "username": "deleted",
        "port": 51826,
        "pin": "333-88-333"
    },

    "platforms": [
        {
            "platform": "Hs100",
            "plugs": [
        {
        "host": "192.168.0.173",
        "port": 9999
        },
        {
        "host": "192.168.0.165",
        "port": 9999
        },
        {
        "host": "192.168.0.193",
        "port": 9999
        }]
        }
    ]
}
```

Homebridge starts but does not do anything after it initialized:

```
> *** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
> *** WARNING *** Please fix your application to use the native API of Avahi!
> *** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node>
> *** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
> *** WARNING *** Please fix your application to use the native API of Avahi!
> *** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node&f=DNSServiceRegister>
> [4/7/2017, 9:05:45 AM] Loaded plugin: homebridge-hs100
> [4/7/2017, 9:05:45 AM] Registering platform 'homebridge-hs100.Hs100'
> [4/7/2017, 9:05:45 AM] ---
> [4/7/2017, 9:05:45 AM] Loaded plugin: homebridge-server
> [4/7/2017, 9:05:45 AM] Registering platform 'homebridge-server.Server'
> [4/7/2017, 9:05:45 AM] ---
> [4/7/2017, 9:05:45 AM] Loaded config.json with 0 accessories and 1 platforms.
> [4/7/2017, 9:05:45 AM] ---
> [4/7/2017, 9:05:45 AM] Loading 1 platforms...
> [4/7/2017, 9:05:45 AM] Initializing Hs100 platform...
> Scan this code with your HomeKit App on your iOS device to pair with Homebridge:
>                        
>     ┌────────────┐     
>     │ 333-88-333 │     
>     └────────────┘     
>                        
> [4/7/2017, 9:05:46 AM] Configuring: Kamerlamp
> [4/7/2017, 9:05:46 AM] Configuring: Autootjeskast
> [4/7/2017, 9:05:46 AM] Configuring: Living Colors Lamp
> [4/7/2017, 9:05:46 AM] Homebridge is running on port 51826.
> [4/7/2017, 9:05:46 AM] Done Configuring Cached Accessories: ["Kamerlamp","Autootjeskast","Living Colors Lamp"]
> [4/7/2017, 9:05:46 AM] Done Configuring Accessories: ["Kamerlamp","Autootjeskast","Living Colors Lamp"]
```
I did downgrade to version 1.1.1 for hs100 because of the bug in the 2.0 as stated by Tucknology here.

any clues? thanks in advance!

**Comment by leematt** [comment]
@MrTeal It's been a while since I've used version 1.1.1, but it looks like everything is being initialized and configured correctly. Once Homebridge is running, it waits for communication with your HomeKit apps. Have you followed the [iOS configuration instructions for Homebridge](https://github.com/nfarina/homebridge#adding-homebridge-to-ios)?

**Comment by MrTeal** [comment]
@leematt thanks! i just needed to re add homebridge to my homekit and now it works again! thanks a lot! the difference with 1.1.1 is that it does not show logging anymore (lamp x "online") so i figured it wasn't working.

**Comment by leematt** [comment]
@MrTeal No problem. Glad you got it working! I've submitted PR #18 to address the problem with 2.0.0, though with a correct `config.json` that version should also work.

**Comment by leematt** [comment]
> I want to make homebridge start at login, don't suppose anyone knows an easy way to do that?

@ghope10 It depends on your OS. Try searching the [Homebridge wiki](https://github.com/nfarina/homebridge/wiki). There are guides for [macOS](https://github.com/nfarina/homebridge/wiki/Install-Homebridge-on-macOS#how-to-automatically-start-homebridge-with-launchd) and [Ubuntu](https://github.com/nfarina/homebridge/wiki/Homebridge-autostart-at-boot-(init.d)-on-Ubuntu-(linux)).

**Comment by quacktacular** [comment]
Anyone have ideas about fixing the dynamic IP issue in 2.0.0? I have it working with a static ip on the hs100, but it would be better to find the actual problem.

**Comment by leematt** [comment]
@quacktacular Aside from deleting the Homebridge accessories cache (`.homebridge/accessories/cachedAccessories`), or manually updating the device IPs in the cache when they change, I'm not aware of a workaround. The plugin probably needs to be updated to recognize the readdressed devices and update the cache accordingly.

**Comment by plasticrake** [comment]
Fixed in 2.0.1 just released. Should now be properly updating to the new IP address when reading from the cache
